src/ast/codegen_llvm.cpp

```diff
@@ -655,87 +657,86 @@ void CodegenLLVM::visit(Call &call)
-    b_.CREATE_MEMSET(str_buf, b_.getInt8(0), map.ident.length() + 1, 1);
-    b_.CreateStore(const_str, str_buf);
-    ArrayType *perfdata_type = ArrayType::get(b_.getInt8Ty(), sizeof(uint64_t) + 2 * sizeof(uint64_t) + map.ident.length() + 1);
-    AllocaInst *perfdata = b_.CreateAllocaBPF(perfdata_type, "perfdata");
+
+    AllocaInst *buf = b_.CreateAllocaBPF(print_struct, "print_" + map.ident);
 
     // store asyncactionid:
-    b_.CreateStore(b_.getInt64(asyncactionint(AsyncAction::print)), perfdata);
+    b_.CreateStore(b_.getInt64(asyncactionint(AsyncAction::print)),
+                   b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(0) }));
 
-    // store top:
-    if (call.vargs->size() > 1)
-    {
-      Integer &top_arg = static_cast<Integer&>(*call.vargs->at(1));
-      Value *top;
-      top_arg.accept(*this);
-      top = expr_;
-      b_.CreateStore(top, b_.CreateGEP(perfdata, {b_.getInt32(0), b_.getInt32(sizeof(uint64_t))}));
-    }
-    else
-      b_.CreateStore(b_.getInt64(0), b_.CreateGEP(perfdata, {b_.getInt64(0), b_.getInt64(sizeof(uint64_t))}));
+    auto id = bpftrace_.maps_[map.ident]->id;
+    auto *ident_ptr = b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(1) });
+    b_.CreateStore(b_.getInt32(id), ident_ptr);
 
-    // store top:
-    if (call.vargs->size() > 2)
+    // top, div
+    for (size_t i = 1; i < 3; i++)
     {
-      Integer &div_arg = static_cast<Integer&>(*call.vargs->at(2));
-      Value *div;
-      div_arg.accept(*this);
-      div = expr_;
-      b_.CreateStore(div, b_.CreateGEP(perfdata, {b_.getInt64(0), b_.getInt64(sizeof(uint64_t) + sizeof(uint64_t))}));
+      llvm::Value *val;
+      if (call.vargs->size() > i)
+      {
+        call.vargs->at(i)->accept(*this);
+        val = b_.CreateIntCast(expr_, elements.at(i), false);
+      }
+      else
+        val = b_.GetIntSameSize(0, elements.at(i));
+
+      b_.CreateStore(val,
+                     b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(i + 1) }));
     }
-    else
-      b_.CreateStore(b_.getInt64(0), b_.CreateGEP(perfdata, {b_.getInt64(0), b_.getInt64(sizeof(uint64_t) + sizeof(uint64_t))}));
 
-    // store map ident:
-    b_.CREATE_MEMCPY(b_.CreateGEP(perfdata, {b_.getInt64(0), b_.getInt64(sizeof(uint64_t) + 2 * sizeof(uint64_t))}), str_buf, map.ident.length() + 1, 1);
-    b_.CreatePerfEventOutput(ctx_, perfdata, sizeof(uint64_t) + 2 * sizeof(uint64_t) + map.ident.length() + 1);
-    b_.CreateLifetimeEnd(perfdata);
+    b_.CreatePerfEventOutput(ctx_, buf, AsyncEvent::Print().LLVMTypeSize());
+    b_.CreateLifetimeEnd(buf);
     expr_ = nullptr;
   }
   else if (call.func == "clear" || call.func == "zero")
   {
+    auto elements = AsyncEvent::MapEvent().asLLVMType(&b_);
+    StructType *print_struct = b_.GetStructType(call.func + "_t",
+                                                elements,
+                                                false);
```

Comment:
these structs should be packed to be able to use `layout_->getAllocSize()` instead of hardcoding the size